odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,716 @@
+        if num_elems != int(num_elems):
+            raise ValueError('trying to read {} bytes, corresponding to '
+                             '{} elements of type {}'
+                             ''.format(dend_abs - dstart_abs, num_elems,
+                                       self.data_dtype))
+        self.file.seek(dstart_abs)
+        # Numpy determines byte order by itself
+        data = np.fromfile(self.file, dtype=self.data_dtype,
+                           count=int(num_elems))
+
+        if dstart_abs == self.header_size and dend_abs == filesize_bytes:
+            # Full dataset read, reshape to stored shape.
+            data = data.reshape(self.data_shape, order=reshape_order)
+
+        return data
+
+
+class FileWriterRawBinaryWithHeader(object):
+
+    """Writer for uncompressed binary files using an optional header.
+
+    This class can be used to write a single binary header followed by a
+    single block of uncompressed binary data to a file.
+    Alternatively, the header can be bypassed and data blocks can be
+    written directly using `write_data`, which allows to write arbitrary
+    portions.
+    """
+
+    def __init__(self, file, header=OrderedDict()):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        file : file-like or str
+            Stream or filename to which to write the data. The stream
+            is allowed to be already opened in a writable mode.
+        header : `OrderedDict`, optional
+            Header in form of an ordered dictionary, where each entry has
+            the following form::
+
+                'name': {'value': value_as_array,
+                         'offset': offset_in_bytes
+                         'description': description_string}
+
+            All ``'value'``'s must be `numpy.ndarray`'s. Their size and the
+            ``'offset'`` entry determine the space that the value occupies
+            in `file`'s header.
+
+            If an empty dictionary is given, no header is written.
+
+        Notes
+        -----
+        **Important:** There is no check if the writes from the provided
+        header are consistent, i.e., that they don't overwrite each other.
+        Ensuring this is the user's responsibility.
+
+        See Also
+        --------
+        header_fields_from_table :
+            Function to parse a specification table, returning a field
+            sequence usable as ``header_fields`` parameter.
+        FileReaderRawBinaryWithHeader.read_header
+        """
+        # Need those attrs in subsequent code
+        file_attrs = ('mode', 'seek', 'write', 'close')
+        is_file = all(hasattr(file, attr) for attr in file_attrs)
+        if is_file:
+            self.__file = file
+            self.__file_shared = True
+        else:
+            self.__file = open(file, 'wb')
+            self.__file_shared = False
+
+        if 'b' not in self.file.mode:
```

Comment:
The file could be opened in `r+` mode which would also be writable. I had a check for `file.is_writable()` before, but temporary files don't have that method. So basically I'm relying on the OS to tell the user whether or not they can write. Not everything has to be double-checked.